### PR TITLE
Updates to VecOps library to use pre-compiled templates for common types

### PR DESCRIPTION
**Pull request by amadio**

### CODE REVIEW

File: math/vecops/inc/ROOT/TVec.hxx [hunk 1]
```diff
@@ -215,7 +216,7 @@ private:
 
 public:
    // ctors
-   TVec() = default;
```
Comment (dpiparo) [inline math/vecops/inc/ROOT/TVec.hxx]: why?

File: math/vecops/inc/ROOT/TVec.hxx [hunk 2]
```diff
@@ -263,12 +264,12 @@ public:
...
-   const_iterator crbegin() const noexcept { return fData.crbegin(); }
-   iterator rend() noexcept { return fData.rend(); }
-   const_iterator rend() const noexcept { return fData.rend(); }
-   const_iterator crend() const noexcept { return fData.crend(); }
+   reverse_iterator rbegin() noexcept { return fData.rbegin(); }
+   const_reverse_iterator rbegin() const noexcept { return fData.rbegin(); }
+   const_reverse_iterator crbegin() const noexcept { return fData.crbegin(); }
+   reverse_iterator rend() noexcept { return fData.rend(); }
+   const_reverse_iterator rend() const noexcept { return fData.rend(); }
+   const_reverse_iterator crend() const noexcept { return fData.crend(); }
```
Comment (dpiparo) [inline math/vecops/inc/ROOT/TVec.hxx]: nice!

File: math/vecops/inc/ROOT/TVec.hxx (hunk 1 above)
Comment (amadio) [reply math/vecops/inc/ROOT/TVec.hxx]: If the constructor is inlined, the symbol is missing in the library.

Comment (bluehood) [reply math/vecops/inc/ROOT/TVec.hxx]: is the ctor not inlined now that it's implicit?

**Comment by dpiparo** [comment]
@phsft-bot  build please

**Comment by bluehood** [comment]
Hi, I think it would be good, if possible, to amend the description of commit "Use extern templates for TVecs of common standard types" with the motivation behind their introduction.
I can see them being accidentally removed and never put back...

File: math/vecops/inc/ROOT/TVec.hxx (hunk 1 above)
Comment (amadio) [reply math/vecops/inc/ROOT/TVec.hxx]: Not for all compilers, unfortunately. Some compilers still inline it. :-/
We may have to add an attribute to force not inlining. But that will affect performance for all types.
Alternatively, we may declare the constructors explicitly, along with the specializations.

**Comment by amadio** [comment]
@bluehood Ok, good point. I will amend the commit message.

**Comment by dpiparo** [comment]
@amadio @bluehood I do not get all these missing symbols errors and retriggered the build. If the PR builds pass, I am in favour to quickly merge so to proceed with the second part of the plan, i.e.:
- Add extern specialisation for all TVec operations
- Compile libVecOps with Ofast
- VDT On by default
- If VDT available, use VDT functions instead of standard ones
all this will be transparent for the user who will just see things boosted...

**Comment by amadio** [comment]
Hi @dpiparo , I agree with the remainder of the plan, but I'd prefer to merge at the end, once the problem of missing symbols is resolved for all platforms. We can merge the first two commits anytime, though, as they do not affect current builds.

**Comment by dpiparo** [comment]
Hi @amadio, I think I found the problem. We simply need to add dependencies on libVecOps.so. For the dependency of PyROOT on VecOps, we can discuss a solution together with @etejedor. I have an idea on how to solve this relying on the fact that a TVec has the same layout of a vector with a custom allocator.

**Comment by bluehood** [comment]
Hi @dpiparo , on this topic #1888 adds libVecOps as a dependency of libTreePlayer.
The failures in that PR should have been fixed by commit 62fca2cbef, so I just restarted the tests to check it's now green.

File: math/vecops/inc/ROOT/TVec.hxx [hunk 3]
```diff
@@ -214,23 +215,30 @@ private:
...
-   TVec(const TVec<T> &) = default;
-   TVec(TVec<T> &&) = default;
-   template <class InputIt>
-   TVec(InputIt first, InputIt last) : fData(first, last)
-   {
-   }
-   TVec(size_type count, const T &value) : fData(count, value) {}
+   // Note: We must define - rather than just declare - all constructors and
+   // operators here so that extern templates have all needed symbols. That is,
+   // we cannot use something like TVec(...) = default; since it only declares.
```
Comment (Axel-Naumann) [inline math/vecops/inc/ROOT/TVec.hxx]: That statement is incorrect; `= default` will work - once I fix the bug in cling! Working on it...

Comment (amadio) [reply math/vecops/inc/ROOT/TVec.hxx]: Hi @Axel-Naumann, Ok, thanks for letting us know. If cling works with the `= default;` way of declaring the constructors, then I will drop this commit before merging.

File: math/vecops/inc/LinkDef.h [hunk 4]
```diff
@@ -10,17 +10,19 @@
...
-#pragma link C++ class ROOT::Experimental::VecOps::TVec<ULong_t>::Impl_t+;
-#pragma link C++ class ROOT::Experimental::VecOps::TVec<ULong64_t>::Impl_t+;
-#pragma link C++ class ROOT::Experimental::VecOps::TVec<UShort_t>::Impl_t+;
-#pragma link C++ class ROOT::Experimental::VecOps::TVec<UChar_t>::Impl_t+;
-#pragma link C++ class ROOT::Experimental::VecOps::TVec<Int_t>::Impl_t+;
-#pragma link C++ class ROOT::Experimental::VecOps::TVec<Long_t>::Impl_t+;
-#pragma link C++ class ROOT::Experimental::VecOps::TVec<Long64_t>::Impl_t+;
-#pragma link C++ class ROOT::Experimental::VecOps::TVec<Short_t>::Impl_t+;
-#pragma link C++ class ROOT::Experimental::VecOps::TVec<Char_t>::Impl_t+;
+#pragma link C++ class ROOT::Experimental::VecOps::TVec<double>::Impl_t+;
```
Comment (Axel-Naumann) [inline math/vecops/inc/LinkDef.h]: Adding `ROOT::Experimental::VecOps::TVec<double>` will enable autoloading of `libVecOps`. This fixes the test failures we see (I tested `./tree/treeplayer/test/dataframe_histomodels`), even without an explicit dependency from `libTreePlayer` to `libVecOps`.

*Both* should happen: users of `TDataFrame` should rely on `libVecOps` being loaded by the linker, *and* autoloading should work.

Comment (Axel-Naumann) [reply math/vecops/inc/LinkDef.h]: See c668e3c

Comment (amadio) [reply math/vecops/inc/LinkDef.h]: Ok, so I have two questions: how to push to an open PR? I never did that before. And also, can we declare loading of nested types and survive with only what you added here?

Comment (bluehood) [reply math/vecops/inc/LinkDef.h]: @amadio regarding your first question, there is a `allow edits from maintainers` checkbox that is probably checked in your case, which allows root maintainers to push to your branch

**Comment by dpiparo** [comment]
@phsft-bot build

**Comment by dpiparo** [comment]
Merged as #1945 adding 2 simple workarounds.

**Pull request closed without merging**